db/compaction/compaction_picker_universal.cc

```diff
@@ -1288,7 +1307,15 @@ Compaction* UniversalCompactionBuilder::PickDeleteTriggeredCompaction() {
         continue;
       }
       FileMetaData* f = vstorage_->LevelFiles(0)[loop];
-      if (f->marked_for_compaction) {
+      // We attempt to involve standalone range tombstone file in compaction
```

Comment:
Yes, this files picking logic doesn't completely prevent prematurely picking such a file. It can end up getting picked into a compaction still. 

I think this logic works good enough for now because, this selection is for when such a file is selected for a start level inputs, that's when a standalone range deletion file is mostly useful for the later optimizations.  The expected atomic replace work flow is rather simple, where there is only an infrequent replace and no live writes, hopefully this picking logic should work to avoid for the majority of the times that standalone range deletion files getting picked prematurely. 

I'm trying to get a working version of this optimization, I can work on more augmentations in follow ups.

Reply:
Since atomic replacement will add at least 2 sorted runs, I think there's a good chance that the compaction picker will choose a compaction to reduce sorted runs, or a space-amp compaction.

Reply:
Are there options to configure a higher threshold to trigger those type of compactions?

Reply:
Yeah, they are configured compactions options. I guess these triggers can be adjusted before atomic replacement.

Reply:
Yeah, I think in this case, what we really want is to tweak the priority of this compaction getting picked. One scenario I can think of is in the atomic data replacement case. We have a sandwiched three levels with the standalone deletion file in between. We can have logic to avoid the standalone deletion file being picked before it matures beyond the oldest snapshot in all cases, but there is not much we can do to prevent it from going into compaction with the upper layer of the sandwich first. That would leave the standalone range deletion file unuseful for optimization even though we tried out best to only pick it up after it is older than the earliest snapshot.

Reply:
> there is not much we can do to prevent it from going into compaction with the upper layer of the sandwich first

We can prevent these compactions in the compaction logic too. My suggestion above was to mark the middle level (with standalone range tombstone) so that it's not picked by other compactions. This is similar to how a sorted run can be marked as being_compacted and is excluded in compaction picking. Then, in PickFilesMarkedForCompaction(), we can exclude certain levels from being picked if their next level is marked as containing standalone range tombstone.

Reply:
Thanks a lot for the detailed implication instructions. I wasn't aware this logic is something that we can implement. Now I have updated the PR to do this instead. I didn't get time to add unit test to verify that size amplification and sorted run compactions will definitely not pick these files. This is fairly no risk and I can add some for this in a follow up. I spent most of my time adding the file filtering thing into stress test. I think making that part work should give good confidence for this feature.